File: frontend/src/utils/chatIds.ts
/** Mirrors Backend app.utils.chat_ids (patient + role composite id). */

export const CHAT_ID_BASE = 100_000_000;

const ROLE_TO_CODE: Record<string, number> = { doctor: 1, caretaker: 2, cho: 3 };

export type ChatRole = 'doctor' | 'caretaker' | 'cho';

export function chatUserId(patientId: number, role: ChatRole): number {
  const code = ROLE_TO_CODE[role];
  if (code === undefined) throw new Error(`Invalid role: ${role}`);
  if (patientId < 0 || patientId >= CHAT_ID_BASE) {
    throw new Error(`patient_id out of range: ${patientId}`);
  }
  return CHAT_ID_BASE * code + patientId;
}

File: Backend/app/utils/chat_ids.py
"""
Stable integer IDs for chat participants: patient context + role.
Format: CHAT_ID_BASE * role_code + patient_id
"""
from __future__ import annotations

CHAT_ID_BASE = 100_000_000

ROLE_TO_CODE: dict[str, int] = {"doctor": 1, "caretaker": 2, "cho": 3}
CODE_TO_ROLE: dict[int, str] = {1: "doctor", 2: "caretaker", 3: "cho"}


def chat_user_id(patient_id: int, role: str) -> int:
    key = role.lower().strip()
    code = ROLE_TO_CODE.get(key)
    if not code:
        raise ValueError(f"Invalid chat role: {role}")
    if patient_id < 0 or patient_id >= CHAT_ID_BASE:
        raise ValueError(f"patient_id out of range: {patient_id}")
    return CHAT_ID_BASE * code + patient_id


def decode_chat_user_id(uid: int) -> tuple[int, str]:
    if uid < 0:
        raise ValueError("Invalid user id")
    code = uid // CHAT_ID_BASE
    patient_id = uid % CHAT_ID_BASE
    role = CODE_TO_ROLE.get(code, "unknown")
    return patient_id, role
